packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -154,7 +154,7 @@ class UpgradeCommandRunner {
       );
     }
     recordState(flutterVersion);
-    await upgradeChannel(flutterVersion);
+    await upgradeChannel(flutterVersion, gitTagVersion.frameworkVersionFor(upstreamRevision));
```

Comment:
Unfortunately, I don't believe this is right either. The `.frameworkVersionFor()` method is confusingly named. It does not return the correct version for the input revision you input, but rather the `GitTagVersion` that it is a method for. And on line 157, I believe that `gitTagVersion` is the current HEAD, or the version that the tool is currently at.

Please see my comments at @RoyARG02's PR for more context: https://github.com/flutter/flutter/pull/69420/files#r515275570. It seems like are both trying to hook into functionality that we don't yet support.

Reply:
@karamage I've updated #69420 as per suggestions from @christopherfujino. Please have a look at it(I've not updated the tests yet).

If you need any clarifications about my approach, or have a better idea on implementing the feature, we can discuss on the chat.

Reply:
@RoyARG02 @christopherfujino Thanks for your suggestion.
I'll look at #69420.

I am trying to understand your approach.